### ISSUE TITLE
tape.jacobian with chain rule cannot compute the gradient correctly

### ISSUE THREAD

**Issue by wangwei-cmd** [tensorflow/tensorflow#56138]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf2.7

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 21.10

### Python version

3.9.7

### CUDA/cuDNN version

11.3.1/8.1.0.77

### Current Behaviour?

```shell
I want to use the jacobian matrix and chain-rule to compute the gradient. When the composed functions are basic arithmetics, the gradient is right. But when the composed function is a convolution, the gradient  calculated by the chain-rule is not consistent with the one outputed by 'tape.gradient'.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

layer = tf.keras.layers.Conv2D(1, 3, activation='relu', padding="same")
x = tf.random.normal([2,7, 7,1])
xx = tf.random.normal([2,7, 7,1])

with tf.GradientTape(persistent=True) as tape1:
    tape1.watch(x)
    y1 = layer(x)
    y2 = layer(y1)
    tape1.watch(y2)
    loss=tf.keras.losses.MeanSquaredError()(y2,xx)

j1=tape1.jacobian(y2,x)
[g_x,g_y2]=tape1.gradient(loss,[x,y2])
j1_vec=tf.reshape(j1,[x.shape[0]*x.shape[1]*x.shape[2]*x.shape[3],
                     y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3]])
g_y2_vec=tf.reshape(g_y2,[y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3],1])
G=tf.matmul(j1_vec,g_y2_vec)
G_x=tf.reshape(G,[x.shape[0],x.shape[1],x.shape[2],x.shape[3]])
e1=G_x-g_x
print('error value of e1:',tf.reduce_max(abs(e1)).numpy())

with tf.GradientTape(persistent=True) as tape2:
    tape2.watch(x)
    y1 = x**3+100
    y2 = tf.math.cos(y1)+10
    tape2.watch(y2)
    loss=tf.keras.losses.MeanSquaredError()(y2,xx)

j2=tape2.jacobian(y2,x)
[g_x,g_y2]=tape2.gradient(loss,[x,y2])
j2_vec=tf.reshape(j2,[x.shape[0]*x.shape[1]*x.shape[2]*x.shape[3],
                     y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3]])
g_y2_vec=tf.reshape(g_y2,[y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3],1])
G=tf.matmul(j2_vec,g_y2_vec)
G_x=tf.reshape(G,[x.shape[0],x.shape[1],x.shape[2],x.shape[3]])
e2=G_x-g_x
print('error value of e2:',tf.reduce_max(abs(e2)).numpy())
```

### Relevant log output

```shell
error value of e1: 0.10533138
error value of e2: 2.3841858e-07
```
</details>

**Comment by tilakrayal** [comment]
@sachinprasadhs  ,
I was able to reproduce the issue in v2.8, v2.9 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/549b80e890d4194c7182466892d2d1e4/56138.ipynb).

**Comment by tomytsai1** [comment]
Hi @wangwei-cmd  ,

The issue is from the wrong order of matmul in the chain rule. Please change these lines:

From
```
j1_vec=tf.reshape(j1,[x.shape[0]*x.shape[1]*x.shape[2]*x.shape[3],
                     y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3]])
g_y2_vec=tf.reshape(g_y2,[y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3],1])
G=tf.matmul(j1_vec,g_y2_vec)
```
To
```
# the Jacobian matrix is first ordered by f and then by x
j1_vec=tf.reshape(j1,[y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3],
                     x.shape[0]*x.shape[1]*x.shape[2]*x.shape[3]])  

# let gradient to be a "row" vector 
g_y2_vec=tf.reshape(g_y2,[1,y2.shape[0]*y2.shape[1]*y2.shape[2]*y2.shape[3]])

# apply the chain rule of  (dg / dy2)  *  (dy2 / dx) , not the other direction
G=tf.matmul(g_y2_vec,j1_vec)
```

In the other example, when the functions are only self-correlated, the Jacobian matrix is diagonal, so the wrong matrix multiplication order didn't affect the results.

Let me know if you have further questions. Thanks.
